rust/primary/src/messages.rs

```diff
@@ -1,189 +1,255 @@
+use crate::error::{DagError, DagResult};
+use crate::primary::Round;
+use config::{Committee, WorkerId};
+use crypto::{Digest, Hash, PublicKey, Signature, SignatureService};
+use ed25519_dalek::Digest as _;
 use ed25519_dalek::Sha512;
 use serde::{Deserialize, Serialize};
-use std::collections::BTreeSet;
-use tokio::sync::mpsc::Sender;
-use tokio::sync::oneshot;
-
-#[cfg(test)]
-#[path = "tests/messages_tests.rs"]
-pub mod messages_tests;
-
-/// This is the data structure the receive re-computes.
-#[derive(Debug, Clone, Serialize, Deserialize)]
-pub struct BlockHeader {
-    /// The author of the block.
-    pub author: NodeID,
-    /// The round number of the block.
-    pub round: RoundNumber,
-    /// The digests of all transactions in that block.
-    pub transactions_digest: WorkersDigests,
-    /// Links to the parents of the block.
-    pub parents: BTreeSet<(NodeID, Digest)>,
-    /// User-defined metadata.
-    pub metadata: Metadata,
-    /// The current instance of the protocol.
-    pub instance_id: SequenceNumber,
+use std::collections::{BTreeMap, BTreeSet, HashSet};
+use std::convert::TryInto;
+use std::fmt;
+
+#[derive(Clone, Serialize, Deserialize, Default)]
+pub struct Header {
+    pub author: PublicKey,
+    pub round: Round,
+    pub payload: BTreeMap<Digest, WorkerId>,
+    pub parents: BTreeSet<Digest>,
+    pub id: Digest,
+    pub signature: Signature,
 }
-pub const GC_DEPTH: u64 = 10;
-
-impl BlockHeader {
-    /// Verify that the block header is well formed.
-    pub fn check(&self, committee: &Committee) -> Result<(), DagError> {
-        dag_ensure!(
-            self.instance_id == committee.instance_id,
-            DagError::BadInstanceId
-        );
-        Ok(())
-    }
 
-    // Checks if a block has some digest as parent
-    pub fn has_parent(&self, digest: &Digest) -> bool {
-        for (_, parent) in &self.parents {
-            if *parent == *digest {
-                return true;
-            }
+impl Header {
+    pub async fn new(
+        author: PublicKey,
+        round: Round,
+        payload: BTreeMap<Digest, WorkerId>,
+        parents: BTreeSet<Digest>,
+        signature_service: &mut SignatureService,
+    ) -> Self {
+        let header = Self {
+            author,
+            round,
+            payload,
+            parents,
+            id: Digest::default(),
+            signature: Signature::default(),
+        };
+        let id = header.digest();
+        let signature = signature_service.request_signature(id.clone()).await;
+        Self {
+            id,
+            signature,
+            ..header
         }
-        false
     }
 
-    // Checks if a block has at least one parent from a set of digests
-    pub fn has_at_least_one_parent(&self, digests: &[Digest]) -> bool {
-        for digest in digests {
-            if self.has_parent(digest) {
-                return true;
-            }
+    pub fn verify(&self, committee: &Committee) -> DagResult<()> {
```

Comment:
do/should/can  we check that the parents are 2f+1 (or have quorum threshold stake?)

Reply:
At this stage we can't because we may not have the parents in the store. We do that later in the core (and use the syncrhonizer to make sure we have the parents)